File: app/backend/src/presentation/variants/presenters/variants_presenter.rs
use crate::application::dto::VariantSummaryDTO;
use crate::presentation::variants::responses::{VariantListItemResponse, VariantListResponse};

pub struct VariantsPresenter;

impl VariantsPresenter {
    fn present_single(dto: VariantSummaryDTO) -> VariantListItemResponse {
        VariantListItemResponse {
            sku_id: dto.sku_id.value().to_string(),
            price: dto.price,
            sale_price: dto.sale_price,
            image: dto.image,
            material: dto.material,
            dimensions: dto.dimensions,
        }
    }

    pub fn present(dtos: Vec<VariantSummaryDTO>) -> VariantListResponse {
        let variants = dtos.into_iter()
            .map(Self::present_single)
            .collect();

        VariantListResponse { variants }
    }
} 

File: app/backend-ts/src/presentation/variants/presenters/variants.presenter.ts
import {
  VariantSummaryDto,
  FindVariantsItemDto,
} from '../../../application/dto';
import {
  VariantSummaryResponse,
  FindVariantsResponse,
  FindVariantsItemResponse,
} from '../responses';

export class VariantsPresenter {
  static toVariantSummaryResponse(
    dto: VariantSummaryDto,
  ): VariantSummaryResponse {
    return new VariantSummaryResponse({
      id: dto.id,
      skuCode: dto.skuCode,
      productId: dto.productId,
      productName: dto.productName,
      name: dto.name,
      currentPrice: dto.currentPrice,
      isInStock: dto.isInStock,
      stockQuantity: dto.stockQuantity,
      colorId: dto.colorId,
      colorName: dto.colorName,
      colorHex: dto.colorHex,
      dimensions: dto.dimensions,
      material: dto.material,
    });
  }

  static toVariantSummaryListResponse(
    dtos: VariantSummaryDto[],
  ): VariantSummaryResponse[] {
    return dtos.map((dto) => this.toVariantSummaryResponse(dto));
  }

  static toFindVariantsResponse(
    dtos: FindVariantsItemDto[],
  ): FindVariantsResponse {
    const variants = dtos.map(
      (dto) =>
        new FindVariantsItemResponse({
          skuId: dto.skuId,
          price: dto.price,
          salePrice: dto.salePrice || undefined,
          image: dto.image || undefined,
          material: dto.material || undefined,
          dimensions: dto.dimensions || undefined,
        }),
    );

    return new FindVariantsResponse({ variants });
  }
}
